### ISSUE TITLE
Data archive through February 28, 2025

### ISSUE THREAD

**Issue by MathewBiddle** [ioos/marine_life_data_network#122]
### Who is requesting this?

Marine Life

### What is being requested?

- Employ workflow currently in-place (manual) to facilitate netCDF package delivery of trajectory datasets (format of which has already been approved by NCEI and finalized) for archive at NCEI
- Prioritized list of deployments will be flagged by ATN DAC Data Coordinator

### What is the requested deadline and why?

2025-02-28

### What is the current status quo (i.e., what happens if this does not get done)?

- data not making it to NCEI thus we are not meeting our PARR requirements.

### What indicates this is done (i.e., how do we know this is complete)?

- [ ] ~~300~~ 30 deployments archived at NCEI

edit: @MathewBiddle edited to reflect the number of deployments (from 300 to 30) pushed to NCEI.

**Comment by MathewBiddle** [comment]
Moved to `In progress` as we should be working on this now.

**Comment by iwensu0313** [comment]
Update: no new datasetse to archive yet since Dec 1, 2024.

**Comment by iwensu0313** [comment]
Update: no new datasets to archive to NCEI yet since Dec 1, 2024.

**Comment by MathewBiddle** [comment]
Any updates on this? I haven't seen any notifications from NCEI that new data have been archived.

**Comment by iwensu0313** [comment]
@mmckinzie hasn't mentioned any new datasets for archival to NCEI (checked the ATN data meeting notes from last week when I was at the conference), so no updates to my knowledge

**Comment by iwensu0313** [comment]
Oh wait it's automatically flowing. Here's the staging link that shows what's flowing: https://stage-ncei-waf.srv.axds.co/atn/test/

Was waiting for feedback or the green light from IOOS Marine Life team I think to move staging to production.

**Comment by mmckinzie** [comment]
@iwensu0313 Do NOT move any of the files from the test directory to the production folder. They are not ready to go. They do not have the correct list of contributors/creators. And some of the NCEI flags in the ADR were not embedded by me.

**Comment by mmckinzie** [comment]
I will follow-up later today with the list of deployments that are ready to move to production.

**Comment by MathewBiddle** [comment]
@mmckinzie per your last update I was expecting more deployments to be pushed to NCEI. I haven't seen those come through. Anything we can do to advance this?

**Comment by mmckinzie** [comment]
@MathewBiddle An issue with the automated NCEI pipeline/archiver tool in the ADR was discovered that was causing problems with file generation for deployments with reused ptt ids that was either resulting in files to fail or the wrong file to be generated, so I have been unable to push any new files to NCEI. Axiom was looking into it and seems they have found a potential solution, but it requires me to make while workable, not necessarily desirable changes to PIs deployment metadata. I'm working on that now so I can generate another batch of test files to confirm the issue has in fact been resolved. If it has, I can resume flagging deployments for release, in theory. 

But, there is also another related issue that was discovered with regards to hooking deployments up to the portal, also potentially related to reused ptt ids, that is causing previously hooked up deployments to be removed from portal if/when a new deployment with the same ptt is hooked up even if the ids are associated with a different projects/species and have different unique deployment ids. This in turn breaks the connection for data/deployments to the ADR for any ids that were unintentionally removed from the portal and if data/deployment are not connected to the portal and data are not linked to the ADR, I cannot autogenerate the NCEI files, which is also preventing me from releasing data. We're still looking into this issue.  Both issues need to be resolved before we can reliably generate the NCEI files.

**Comment by MathewBiddle** [comment]
Thanks @mmckinzie ! Hopefully those issues can be sorted out and we can continue pushing data to NCEI. 

@iwensu0313 any idea on LOE for these fixes?

**Comment by iwensu0313** [comment]
Not sure what LOE stands for but this would likely be a med/high effort as the source of the issue is unknown so would require extensive troubleshooting, planning, then fixing, then testing. Given the complexity of the requirements and connectivities, the best way to approach this would be to make sure we're addressing all concerns and end requirements instead of one-off fixes. Will share my ATN summary doc later today so we can decide how to move forward over this next bit

**Comment by mmckinzie** [comment]
ADR archiver tool logic improvements along with listing the unique manufacturer/vendor ID in the deployment ID field instead of a unique PI provided ID appears to have resolved the issue with generating NCEI files deployments with reused ptt IDs. 

All 29 files from ADR project 32 are now being generated correctly. OK to move these files from the staging WAF test directory to the prod folder for harvesting by NCEI. 

Likely also OK to go ahead and updated the ADR archiver tool pathway to send NCEI files directly to the prod folder instead of the staging WAF test directory.

**Comment by mmckinzie** [comment]
@MathewBiddle As mentioned above recent improvements to the ADR archiver tool appear to have resolved the issue with file generation for deployments with reused ptt ids. Although we still haven't figure out what is causing the related issue in the portal. 

Steps still need to be taken to fully automate this pathway as generated files are presently sent to the staging WAF test directory and not the production WAF prod directory for harvesting by NCEI. But I think we are OK to go ahead and make that final change to the pipeline. But that work would obviously need to be completed under the March 1+ SOW.

29 deployments are ready to be moved over from test directory for prod directory.

**Comment by MathewBiddle** [comment]
@iwensu0313 can Axiom move those over from test to prod?

To note, 29 << 300. We agreed that the automated process wouldn't be functioning at this point and it would be manual. I guess that's because of the tool that builds the netCDF files was broken? Or, was that a miss on what could be doable given your available time @mmckinzie ?

**Comment by iwensu0313** [comment]
Not sure if this helps answer your question Matt, but here's a recap of what was done: 
- Axiom finished the first iteration of the pipeline that auto-generates NCEI archive files in Fall 2024; the outputs were being sent to a staging WAF so that we could do proper testing w/ Megan and make adjustments as necessary
- We received some feedback from Megan during the Dec 2024 - Feb 2025 period and we implemented some adjustments to the pipeline in Feb 2025 to account for some discrepancies Megan noticed in reviewing the files in staging WAF
- Sounds like Megan is happy with the pipeline, but while we are in this 'service pause' we will wait to point the pipeline to production WAF until the funds we were expecting March 1 2025 come in, as there are some additional things to sort out that would be medium effort
    - Update pipeline to populate auto-generated archive files in production WAF
    - "Send to NCEI" checkbox in ADR deployment form is currently available to anyone with access to ADR including PIs, which Megan has expressed is not desired. Add'l work would have to be done to refine this workflow (e.g. retain this ability only for Admin users aka Megan)

**Comment by iwensu0313** [comment]
Yes,  we can go ahead and move those files Megan approved in staging WAF to production WAF! That is easy and straightforward.

**Comment by mmckinzie** [comment]
Thanks for the recap @iwensu0313 

@MathewBiddle Only 29 deployments were flagged for release due to the additional effort that was required to troubleshoot and correct the issues that were discover with the ADR archiver tool during the Dec 2024-Feb 2025 period after the first iteration of the pipeline was finished late last fall. The implemented solution also required me to make additional updates to all deployments metadata before I can flag files for release, which of course requires extra time/effort. 

As Iwen mentioned above, I believe those issues have been resolved and as of today if is OK for me to resume flagging deployments for release to NCEI. Although in the near-term Axiom will still need to manually shift files from the staging WAF test directory to the production folder for harvesting by NCEI.

**Comment by iwensu0313** [comment]
@mmckinzie did you want us to also clear the staging WAF so there is a clean slate in there https://stage-ncei-waf.srv.axds.co/atn/test/

**Comment by MathewBiddle** [comment]
Looks like two deployments were picked up at NCEI. They need to update some mapping tables on NCEI's end, then they should go through. 

The two packages are listed below.

```
53644-20100927
53631-20100927
```

**Comment by MathewBiddle** [comment]
NCEI only picked up those two because they ran into some issues with the mapping table, so they cancelled the remaining pull.

@mmckinzie @iwensu0313 thank you both for the summaries above. For now, I think ATN has done it's part wrt pushing data to NCEI. While it was a miss on the number of deployments, now we know where some of the gotchas are.

Something to keep in the back of our minds is the Level of Effort (LOE) required to maintain this pathway. If we're constantly updating it with nuanced changes, that will require a lot of effort. We should be shooting at something that is sustainable while providing some level of granularity, but not soo specific that we need to keep fiddling with the codebase for each nuanced case. This is a fine line to walk, but I'm sure we can get there.

**Comment by MathewBiddle** [comment]
I updated the description to reference the 30 deployments to NCEI.

**Comment by mmckinzie** [comment]
@iwensu0313 Yes, please go ahead and clear the staging WAF. Thanks.

**Comment by mmckinzie** [comment]
@MathewBiddle Have you received any updates from NCEI regarding the documentation/mapping table issue that was discovered? 

I want to be able to flag additional deployments for release, but I want to wait until we're sure lingering issues have been resolved and the last batch of files has been successfully integrated into the ATN collection.

**Comment by MathewBiddle** [comment]
I've reached out to John. Please continue to push additional deployments to the WAF. The issues are on NCEI's side and should be resolved soon.

**Comment by iwensu0313** [comment]
@mmckinzie - just pinged team to clear the staging WAF. should be able to do that today or monday
@MathewBiddle did you want to close this and move new updates to https://github.com/ioos/marine_life_data_network/issues/135

**Comment by mmckinzie** [comment]
OK, I will start working on the next batch of deployments.

**Comment by MathewBiddle** [comment]
Thanks all! I'd like to close this once we have NCEI Accession numbers for the packages we intended on submitting. To me, this still is not complete.

**Comment by MathewBiddle** [comment]
This last batch just went through! There are now 40 records at NCEI, with 24 records archived from this last batch.

Congrats! I'm trying to compile the list of accession numbers and will update this with those identifiers.

Closing.

**Comment by MathewBiddle** [comment]
![Image](https://github.com/user-attachments/assets/d595793b-8940-4ff5-b50f-cefc23c6f06f)

**Comment by mmckinzie** [comment]
@MathewBiddle There should be 29 records from this last batch, not 24.

**Comment by mmckinzie** [comment]
The DOI and accession numbers have been added to the ADR, and the list of deployments has been added to the Google Sheet archival tracking document. 

We are still waiting for data from five more deployments to be added to the NCEI ATN collection. The files are in the prod folder.

**Comment by mmckinzie** [comment]
Yeah, I saw that as well. I added the accession numbers for those two deployments to the ADR and updated the Google [sheet](https://docs.google.com/spreadsheets/d/1XDJRza6RPqdLlEep8mmCGYqFynuj-erx/edit?usp=sharing&ouid=113417505623695767850&rtpof=true&sd=true). 

I keep an eye out for the remaining three.

**Comment by mmckinzie** [comment]
The remaining three deployments have been archived at NCEI. The accession numbers have been added to the ADR and the Google sheet, completing this effort.

https://www.ncei.noaa.gov/archive/accession/0303464
https://www.ncei.noaa.gov/archive/accession/0303525
https://www.ncei.noaa.gov/archive/accession/0303526

Note: A Minor issue with the citation for 0303464 was noticed (i.e., missing PI), NCEI was notified, and the issue was corrected.
